File: bin/online-graph.rb
#!/usr/bin/env ruby

require 'priority_queue'
require 'bindata'
require './bin/tweet.rb'
require 'set'

# Do not produce wrong results using the wrong interpreter
throw "Require at least 2.2" if RUBY_VERSION =~ /^(1[.]|2[.][01])/

WINDOW = 60

class Processor
  attr_accessor :latest
  def initialize
    @latest = 0
    @edges = {}
    @queue = PriorityQueue.new()
  end

  def add_edge(ctime, edge)
    edge_key = edge.join(' ')
    # does the edge already exist?
    old_ctime = @edges[edge_key]
    if old_ctime.nil? || ctime > old_ctime
      # update the priority if not exist, or larger.
      @queue[edge_key] = ctime
      @edges[edge_key] = ctime
    end
  end

  def process!(ctime, nodes)
    # ignore if too old.
    return if (@latest - ctime) > WINDOW

    # Update the latest if it is a new tweet.
    if ctime > @latest
      @latest = ctime
    end
    # GC should be called _before_ addition of
    # new nodes.
    gc()
    nodes.combination(2).each do |edge|
      add_edge(ctime, edge)
    end
  end

  def gc_complete?()
    minkval = @queue.min
    return true unless minkval
    mink,minv = minkval
    v = (@latest - minv) > WINDOW
    return !v
  end

  def gc()
    while !gc_complete?
      min_edge, prio = @queue.min
      @queue.delete(min_edge)
      @edges.delete(min_edge)
    end
  end

  def avg()
    nodes = Set.new
    @edges.keys.each do |l_r|
      nodes.merge(l_r.split(' '))
    end
    return (2.0 * @edges.length)/nodes.length
  end
end

def binread
  while !STDIN.eof? do
    t = Tweet.read(STDIN)
    yield t.created_at, t.nodes.to_a
  end
end

def textread
  STDIN.each do |l|
    case l
    when /^@/   # For debug.
      puts l,"\n"
      next
    when /^#/
      next
    end
    created, *nodes = l.split(/,/).map(&:to_i)
    yield created, nodes
  end
end

$binary = if ARGV[0] == '-a' then false else true end

$current = Processor.new
if $binary
  binread do |created, nodes|
    $current.process!(created, nodes)
    puts $current.avg
  end
else
  textread do |created, nodes|
    $current.process!(created, nodes)
    puts $current.avg
  end
end

File: bin/online-graph.py
#!/usr/bin/env python3
import sys
import heapdict
import itertools

#heapq

WINDOW = 60

class Processor:
    def __init__(self):
        self.latest = 0
        self.edges = {}
        self.queue = heapdict.heapdict()

    def add_edge(self, ctime, edge):
        edge_key = ' '.join(map(str,edge))
        old_ctime = self.edges.get(edge_key, None)
        if (not old_ctime) or (ctime > old_ctime):
            self.queue[edge_key] = ctime
            self.edges[edge_key] = ctime

    def process(self, ctime, nodes):
        if (self.latest - ctime ) > WINDOW: return

        if ctime > self.latest:
            self.latest = ctime

        self.gc()

        for edge in itertools.combinations(nodes, 2):
            self.add_edge(ctime, edge)

    def gc_complete(self):
        if len(self.queue) == 0: return True
        edge, ctime = self.queue.peekitem()
        v = (self.latest - ctime) > WINDOW
        return (not v)

    def gc(self):
        while not self.gc_complete():
            min_edge, ctime = self.queue.popitem()
            del self.edges[min_edge]

    def avg(self):
        nodes = set()
        for l_r in self.edges.keys():
            nodes.update(l_r.split(' '))
        return (2.0 * len(self.edges))/len(nodes)


def textread():
    for l in sys.stdin:
        created, *nodes = map(int, l.rstrip('\n').split(','))
        yield (created, nodes)



binary = False if (len(sys.argv) > 1 and sys.argv[1] == '-a') else True

current = Processor()
if binary:
    pass
else:
    for (created, nodes) in textread():
        current.process(created, nodes)
        print(current.avg())
    pass

